File: examples/4.3.0.0/navigate/android/Navigation/app/src/main/java/com/here/navigation/LocationProviderImplementation.java
package com.here.navigation;

import android.content.Context;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.util.Log;

import com.here.sdk.core.GeoCoordinates;
import com.here.sdk.core.Location;
import com.here.sdk.core.errors.InstantiationErrorException;
import com.here.sdk.navigation.LocationListener;
import com.here.sdk.navigation.LocationProvider;
import com.here.sdk.navigation.LocationSimulator;
import com.here.sdk.navigation.LocationSimulatorOptions;
import com.here.sdk.routing.Route;

import java.util.Date;

// A class that conforms the HERE SDK's LocationProvider interface.
// This class is required by Navigator to receive location updates from either the device or the LocationSimulator.
public class LocationProviderImplementation implements LocationProvider {

    private static final String TAG = LocationProviderImplementation.class.getName();

    public static final int TIMEOUT_POLL_INTERVAL_IN_MS = 500;
    public static final int TIMEOUT_INTERVAL_IN_MS = 3000;

    // Set by the Navigator instance to listen for location updates.
    @Nullable
    private LocationListener locationListener;
    @Nullable
    public Location lastKnownLocation;

    // A loop to check for timeouts between location events.
    private final Handler timeoutHandler = new Handler();
    private final PlatformPositioningProvider platformPositioningProvider;
    private LocationSimulator locationSimulator;
    private boolean isSimulated;

    public LocationProviderImplementation(Context context) {
        platformPositioningProvider = new PlatformPositioningProvider(context);
    }

    // Provides location updates based on the given route.
    public void enableRoutePlayback(Route route) {
        if (locationSimulator != null) {
            locationSimulator.stop();
        }

        locationSimulator = createLocationSimulator(route);
        locationSimulator.start();
        isSimulated = true;
    }

    // Provides location updates based on the device's GPS sensor.
    public void enableDevicePositioning() {
        if (locationSimulator != null) {
            locationSimulator.stop();
            locationSimulator = null;
        }

        isSimulated = false;
    }

    @Override
    public void start() {
        platformPositioningProvider.startLocating(new PlatformPositioningProvider.PlatformLocationListener() {
            @Override
            public void onLocationUpdated(android.location.Location location) {
                if (!isSimulated) {
                    handleLocationUpdate(convertLocation(location));
                }
            }
        });

        timeoutHandler.postDelayed(timeoutRunnable, TIMEOUT_POLL_INTERVAL_IN_MS);
    }

    @Override
    public void stop() {
        platformPositioningProvider.stopLocating();
        timeoutHandler.removeCallbacks(timeoutRunnable);
    }

    @Nullable
    @Override
    public LocationListener getListener() {
        return locationListener;
    }

    @Override
    public void setListener(@Nullable LocationListener locationListener) {
        this.locationListener = locationListener;
    }

    private void handleLocationUpdate(Location location) {
        lastKnownLocation = location;

        if (locationListener != null) {
            // The GPS location we received from either the platform or the LocationSimulator is forwarded to the Navigator.
            locationListener.onLocationUpdated(location);
        }
    }

    private final Runnable timeoutRunnable = new Runnable(){
        public void run(){
            try {
                timeoutLoop();
                timeoutHandler.postDelayed(this, TIMEOUT_POLL_INTERVAL_IN_MS);
            } catch (Exception exception) {
                exception.printStackTrace();
            }
        }
    };

    private void timeoutLoop() {
        if (isSimulated) {
            // LocationSimulator already includes simulated timeout events.
            return;
        }

        if (lastKnownLocation != null) {
            double millisecondsSince1970 = lastKnownLocation.timestamp.getTime();
            double timeIntervalInMs = System.currentTimeMillis() - millisecondsSince1970;
            if (locationListener != null && timeIntervalInMs > TIMEOUT_INTERVAL_IN_MS) {
                //If the last location is older than TIMEOUT_INTERVAL_IN_MS, we forward a timeout event to Navigator.
                locationListener.onLocationTimeout();
                Log.d(TAG, "GPS timeout detected: " + timeIntervalInMs);
            }
        }
    }

    // Provides fake GPS signals based on the route geometry.
    // LocationSimulator can also be set directly to the Navigator, but here we want to have the flexibility to
    // switch between real and simulated GPS data.
    private LocationSimulator createLocationSimulator(Route route) {
        double speedFactor = 10;
        int notificationIntervalInMilliseconds = 100;
        LocationSimulatorOptions locationSimulatorOptions =
                new LocationSimulatorOptions(speedFactor, notificationIntervalInMilliseconds);

        LocationSimulator locationSimulator;

        try {
            locationSimulator = new LocationSimulator(route, locationSimulatorOptions);
        } catch (InstantiationErrorException e) {
            throw new RuntimeException("Initialization of LocationSimulator failed: " + e.error.name());
        }

        locationSimulator.setListener(new LocationListener() {
            @Override
            public void onLocationUpdated(Location location) {
                if (isSimulated) {
                    handleLocationUpdate(location);
                }
            }

            @Override
            public void onLocationTimeout() {
                if (isSimulated) {
                    locationListener.onLocationTimeout();
                }
            }
        });

        return locationSimulator;
    }

    // Converts platform location to com.here.sdk.core.Location.
    private Location convertLocation(android.location.Location nativeLocation) {
        GeoCoordinates geoCoordinates = new GeoCoordinates(
                nativeLocation.getLatitude(),
                nativeLocation.getLongitude(),
                nativeLocation.getAltitude());
        Location location = new Location(geoCoordinates, new Date());
        location.bearingInDegrees = (double) nativeLocation.getBearing();
        location.speedInMetersPerSecond = (double) nativeLocation.getSpeed();
        location.horizontalAccuracyInMeters = (double) nativeLocation.getAccuracy();

        return location;
    }
}

File: examples/other_releases/4.3.0.0/navigate/ios/Navigation/Navigation/LocationProviderImplementation.swift
/*
 * Copyright (C) 2019-2020 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import CoreLocation
import heresdk
import UIKit

// A class that conforms the HERE SDK's LocationProvider protocol.
// This class is required by the Navigator to receive location updates from either the device or the LocationSimulator.
class LocationProviderImplementation : LocationProvider,
                                       LocationDelegate,
                                       PlatformPositioningProviderDelegate {

    // Conforms to the LocationProvider protocol.
    // Set by the Navigator instance to listen to location updates.
    var delegate: LocationDelegate?

    var lastKnownLocation: Location?
    private let platformPositioningProvider: PlatformPositioningProvider
    private var locationSimulator: LocationSimulator?
    private var isSimulated: Bool = false

    // A loop to check for timeouts between location events.
    private lazy var timeoutDisplayLink: CADisplayLink = {
        let displayLink = CADisplayLink(target: self,
                                        selector: #selector(timeoutLoop))
        displayLink.preferredFramesPerSecond = 2
        displayLink.add(to: .current, forMode: .common)
        return displayLink
    }()

    init() {
        platformPositioningProvider = PlatformPositioningProvider()
        platformPositioningProvider.delegate = self
    }

    // Provides location updates based on the given route.
    func enableRoutePlayback(route: Route) {
        if let locationSimulator = locationSimulator {
            locationSimulator.stop()
        }

        locationSimulator = createLocationSimulator(route: route)
        locationSimulator!.start()
        isSimulated = true;
    }

    // Provides location updates based on the device's GPS sensor.
    func enableDevicePositioning() {
        if locationSimulator != nil {
            locationSimulator!.stop()
            locationSimulator = nil
        }

        isSimulated = false;
    }

    // Conforms to the LocationProvider protocol.
    func start() {
        platformPositioningProvider.startLocating()
        timeoutDisplayLink.isPaused = false
    }

    // Conforms to the PlatformPositioningProviderDelegate to receive platform GPS events.
    func onLocationUpdated(location: CLLocation) {
        if !isSimulated {
            handleLocationUpdate(location: convertLocation(nativeLocation: location))
        }
    }

    // Conforms to the LocationProvider protocol.
    func stop() {
        platformPositioningProvider.stopLocating()
        timeoutDisplayLink.isPaused = true
    }

    private func handleLocationUpdate(location: Location) {
        // The GPS location we received from either the platform or the LocationSimulator is forwarded to the Navigator.
        delegate?.onLocationUpdated(location)
        lastKnownLocation = location
    }

    @objc private func timeoutLoop() {
        if isSimulated {
            // LocationSimulator already includes simulated timeout events.
            return
        }

        if let lastKnownLocation = lastKnownLocation {
            let timeIntervalInSeconds = lastKnownLocation.timestamp.timeIntervalSinceNow * -1
            if timeIntervalInSeconds > 3 {
                //If last location is older than 3 seconds we forward a timeout event to Navigator.
                delegate?.onLocationTimeout()
                print("GPS timeout detected: \(timeIntervalInSeconds)")
            }
        }
    }

    // Provides fake GPS signals based on the route geometry.
    // LocationSimulator can also be set directly to the Navigator, but here we want to have the flexibility to
    // switch between real and simulated GPS data.
    private func createLocationSimulator(route: Route) -> LocationSimulator {
        let locationSimulatorOptions = LocationSimulatorOptions(speedFactor: 10,
                                                                notificationIntervalInMilliseconds: 100)
        let locationSimulator: LocationSimulator

        do {
            try locationSimulator = LocationSimulator(route: route,
                                                      options: locationSimulatorOptions)
        } catch let instantiationError {
            fatalError("Failed to initialize LocationSimulator. Cause: \(instantiationError)")
        }

        locationSimulator.delegate = self
        locationSimulator.start()

        return locationSimulator
    }

    // Conforms to the LocationDelegate, which is required to send notifications from the LocationSimulator.
    func onLocationUpdated(_ location: Location) {
        if isSimulated {
            handleLocationUpdate(location: location)
        }
    }

    // Conforms to the LocationDelegate, which is required to send notifications from the LocationSimulator.
    func onLocationTimeout() {
        if isSimulated {
            delegate?.onLocationTimeout()
        }
    }

    // Converts platform CLLocation to HERE SDK Location.
    private func convertLocation(nativeLocation: CLLocation) -> Location {
        let geoCoordinates = GeoCoordinates(latitude: nativeLocation.coordinate.latitude,
                                            longitude: nativeLocation.coordinate.longitude,
                                            altitude: nativeLocation.altitude)
        var location = Location(coordinates: geoCoordinates,
                                timestamp: nativeLocation.timestamp)
        location.bearingInDegrees = nativeLocation.course
        location.speedInMetersPerSecond = nativeLocation.speed
        location.horizontalAccuracyInMeters = nativeLocation.horizontalAccuracy
        location.verticalAccuracyInMeters = nativeLocation.verticalAccuracy

        return location
    }
}
